### ISSUE TITLE
git-secret will always prefer gnupg1 over gnupg2 on fedora 29

### ISSUE THREAD

**Issue by joshrabinowitz** [sobolevn/git-secret#426]
#### Issue:

On fedora 29 (where git-secret is now in 'stable'), the distribution packages both `gnupg` (which is gpg v1), and `gnupg2` (which is unsurprisingly gpg v2).

Generally speaking it would be preferable if git-secret defaulted to using gnupg2, but it currently does the opposite:

* the specfile `Requires:       gpg         >= 1.4` (See https://pagure.io/git-secret/raw/master/f/git-secret.spec). If the `gnupg` package isn't installed when you run `dnf install git-secret` (even if `gnupg2` is installed) then it is installed with `git-secret`

* If the user installs both the `gnupg` and `gnupg2` packages, the system will have both `/usr/bin/gpg` and `/usr/bin/gpg2`. By default, at runtime, git-secret will _always_ default to the `gpg` (gnupg version 1) binary.

* To use gnupg2 on a fedora 29 system a user must set the env var `SECRETS_GPG_COMMAND` to `gpg2`.

Changing the default version of gpg used can be a little thorny because we don't want to switch which version of gpg we're using out from under users who are already using `git-secret`, but we do probably want to depend on, and default to, using the `gnupg2` package and `gpg2` binary on fedora.

This is also related to #208 and https://github.com/sobolevn/git-secret/blob/master/RFC/RFC001.md because gpg & git-secret setups created with one version of gnupg may not interoperate correctly with ones made from different versions of gnupg.

#### Suggested Fixes:

* The git-secret `.specfile` should depend on `gnupg2` and not `gnupg` in Fedora, and 

* git-secret should default to looking for `/usr/bin/gpg2` and if it exists using that over `/usr/bin/gpg` (This behavior would be overridden by using the `SECRETS_GPG_COMMAND` env var) (NOTE: This is a big enough 'interface' change that it definitely warrants bumping the version to 0.3.0)

* It would be ideal if git-secret had a way of recording or detecting (roughly, or exactly) which version of gnupg it used to create keychains, and could warn the user if problems might occur. (I do recognize Here Be Dragons).

* If we knew which version of gnupg was used to create keychains, we might be able to choose the gpg or gpg2 binary to match (or at least not _known_ to potentially conflict)

**Comment by joshrabinowitz** [comment]
@gombosg let us know if you have any thoughts about the `specfile` item mentioned above

**Comment by gombosg** [comment]
Sure. During specfile creation I just followed the readme.md saying "gpg since gnupg 1.4 to gnupg 2.X".
I ran into the same issue when trying to sign commits with Git. I had to manually set
```
[gpg]
	program = gpg2
```
in `.gitconfig` for the signing to work.

Note that we can have [boolean dependencies](http://rpm.org/user_doc/boolean_dependencies.html) and [weak dependencies](https://rpm.org/user_doc/dependencies.html#weak-dependencies) (recommends), too.

Recommends means it will install the package if possible.

My question: how do gnupg 1.x and 2.x interoperate? If a user started with 1.x for git-secret but switches to 2.x later, are they totally compatible?

I Suggest pushing the specfile update along with the git-secret version update so that it always handles these versions correctly.

I guess this is all similar to Python 3... some breaking changes...

**Comment by joshrabinowitz** [comment]
I think we're going to have to close this as a `wontfix`. Here's why.

* This isn't just a fedora related issue. The code is currently written to always look for a 'gpg' binary, so whatever version of gpg that binary is, that's what git-secret has been happily using all along. Changing this default behavior seems dangerous.

* gpg1 and gpg2 aren't particularly interoperable and can cause problems working together. This is discussed in #136 and #208  . 

* On some 'modern' systems, gnupg2 installs the same binary as `gpg` and `gpg2`. On these systems git-secret is already using gnupg2 (via the `gpg` binary) and no change is needed.

* All the above means that if we change the default behavior (without care), some git-secret setups that currently work might break. We could do a bunch of 'look at their git-secret setup and see if we can figure out which installed binary of gpg/gpg2 is the right one' but this seems like a dead end when we ultimately want to move towards the RFC described in #208, which aims to remove dependence on the particular version of gpg used for gig-secret operation.

Let me know if you have more thoughts! Planning to close this issue as #wontfix soon.



This means that if we change the default behavior (without care), some git-secret setups that currently work might break.

**Comment by gombosg** [comment]
Okay I still need to change the specfile since gpg2 is provided by a different package (gnupg2 and not gnupg).

I can set it to 
```
Requires: (gnupg2 or gnupg)
```
so that GPG2 is preferred as a default if you wish.

**Comment by joshrabinowitz** [comment]
@gombosg , I just tested on fedora 29, and there are problems with preferring `gnupg2` over `gnupg` 

On Fedora 29,
*  `/usr/bin/gpg` (version 1.4.23) is only provided by the `gnupg` rpm, while 
* `gnupg2` installs `/usr/bin/gpg2` (version 2.2.13), which by default `git-secret` will not find, as it'll only look for `gpg`.

To re-iterate, currently, unless the user sets the SECRETS_GPG_COMMAND env var, `git-secret` will not find version 2 of gpg on fedora 29.

This issue doesn't exist on CentOS 7 (and I presume RH7), which doesn't provide a `gnupg` package. On Centos7 `gnupg2` installs version `2.0.22 `in both `/usr/bin/gpg` and `/usr/bin/gpg2` (Although note the contemporary 2019 version of gpg is 2.2.*, not 2.0.*)

Question:

* Why does fedora 29 package gnupg1 and gnupg2 (especially when CentOS 7 does not package gnupg1)?

Solution Ideas:

1) One solution to this is to have the next release, version 0.3.0, default to preferring `gpg2` over `gpg2` when possible, and documenting that widely. 
I'd really prefer to have something like https://github.com/sobolevn/git-secret/blob/master/RFC/RFC001.md implemented before we do that, but on the other hand, it's not ideal to just leave new users using gnupg1 when gnupg2 is probably preferable. So I'm almost on the fence about defaulting to using `gpg2` if found, but ultimately am against it because think that it's too big of a change to drop in since it will very likely break some users' setups.

2) Another solution, which I don't love but I think is viable, is looking at the keychain in a given git-secret setup to infer which of the versions of gpg*  _installed on that system_ seems most likely to have created that setup. I suspect we can reasonably detect setups that have been created with gnupg 1.0.* vs 2.0.* vs 2.2.*, and use the correct gpg/gpg2 binary on that system to match. It's not pretty but it keeps everything operating as expected for existing users.

Wrapping up,
* I'm not sure we can default to depending on gnupg2 realistically, and
* I think the way forward is for someone to implement 2), and then for _new_ `git-secret init`setups we could default to using the modern `gpg2` where appropriate. 
* In the long run we should definitely implement something like RFC001 which removes dependence on the binary formats various versions of gpg use. This is the best path forward out of these gpg version conundrums.
